[WIP] Bitbucket API SCM resolution

theckman opened the pull request:
On a plane, so testing this is really hard and it would suck if a "planefactor"
issue made this code disappear. :P -- Will amend this commit message to link to
the related issue once things are looking good (in terms of CI).

This change introduces a new SCM resolution mechanism for Bitbucket repos, to
hopefully avoid false-positives when reaching out to Bitbucket to get the repo's
metadata. Bitbucket is a weird code repository solution in that it supports both
`git` and `hg` (mercurial) for hosting repos. This makes checking out source
code from Bitbucket less straightforward than other places.

Previously the deducer tried to make educated guesses around which SCM was in
use based on a few things, largely hints from the package's import path. It also
had some fallbacks in place in case it got it wrong, but it did still cause some
issues in a few cases when it got it wrong.

This change introduces a minimal internal Bitbucket API client (`bitbucketapi`)
with the single purpose of reaching out to Bitbucket to get the SCM type for the
repository. This deprecates the need for the earlier heuristics, and should
result in a correct result every time. This client was written to be consumed as
a singleton (calling functions on the package) to avoid complicating the
`bitbucketDeducer` further.

The client has no caching of previously seen SCM results for a package, which
could result in more API calls than we'd like. One optimization would be to keep
a map of previously resolved packages, and to serve the result from there.

Tests are also implemented for the `bitbucketapi` package, hitting `97.0%` code
coverage.

Signed-off-by: Tim Heckman <t@heckman.io>

kevinburke commented on gps/deduce.go:
```diff
@@ -167,18 +168,49 @@ func (m bitbucketDeducer) deduceRoot(path string) (string, error) {
 	return "bitbucket.org" + v[2], nil
 }
 
+func getBitbucketRepo(s string) (string, error) {
```
Worth posting a doc string that has a sample input and output?

theckman replied:
I was planning on writing a test for this, since it makes sure it works and accomplishes the same.

kevinburke commented on gps/deduce.go:
```diff
@@ -167,18 +168,49 @@ func (m bitbucketDeducer) deduceRoot(path string) (string, error) {
+
+	repoStr := ss[1] + "/" + ss[2]
+
+	if repoStr[len(repoStr)-4:] == ".git" {
```
`strings.HasSuffix` here?

kevinburke commented on gps/deduce.go:
```diff
@@ -167,18 +168,49 @@ func (m bitbucketDeducer) deduceRoot(path string) (string, error) {
 func (m bitbucketDeducer) deduceSource(path string, u *url.URL) (maybeSources, error) {
 	v := m.regexp.FindStringSubmatch(path)
 	if v == nil {
 		return nil, fmt.Errorf("%s is not a valid path for a source on bitbucket.org", path)
```
should we do `%q` here? in case the empty string gets passed in, it can help clarify what is actually happening.

theckman replied:
Let me take a look at the callers of this to see if that's a potential scenario. If so it wouldn't hurt, although I'd likely raise it as its own PR just to avoid too much change in this one.

kevinburke commented on gps/deduce.go:
```diff
@@ -167,18 +168,49 @@ func (m bitbucketDeducer) deduceRoot(path string) (string, error) {
+		return nil, errors.Wrap(err, "failed to extract repo from bitbucket URL")
+	}
+
+	scm, err := bitbucketapi.GetSCM(bbRepo)
```
`context.Context` here?

theckman commented on gps/internal/bitbucketapi/bitbucketapi.go:
```diff
@@ -0,0 +1,139 @@
+func packageInitClient() {
+	token := os.Getenv(environmentVariable)
+
+	defaultClient = New(pooledClent(), token)
```
LOL!

theckman commented:
@kevinburke @jmank88 I've removed the WIP header from this PR and removed the same info from the commit message. I think this is ready for final review, with the one thing worth probably calling out being this assumption:

- https://github.com/golang/dep/pull/2096/files#diff-b199297aa91ec67eeda3e43e327fef81R214

I'm happy to switch it to a `context.TODO`, but it seems like this is a pretty safe default for the type of API call we're making. If this can't complete within 60 seconds, the person running `dep` is likely going to have a much larger issue.

kevinburke commented on gps/deduce.go:
```diff
@@ -167,36 +169,74 @@ func (m bitbucketDeducer) deduceRoot(path string) (string, error) {
+
+	isgit := scm == "git"
+
 	u.Host = "bitbucket.org"
```
It doesn't seem super great to be manipulating the input?

kevinburke replied:
I guess this is already here

theckman replied:
I'm keeping the scope of this PR to only fix the things I know are broken. We can look at that, and the other thing you raised above about quoting the error string, but they are beyond the scope of this PR.

kevinburke replied:
ok

kevinburke commented on gps/internal/bitbucketapi/bitbucketapi_test.go:
```diff
@@ -0,0 +1,542 @@
+	return mux
+}
+
+func pooledTestClent() *http.Client {
```
Client*

theckman replied:
Fixed.

kevinburke commented on gps/deduce.go:
```diff
@@ -167,36 +169,74 @@ func (m bitbucketDeducer) deduceRoot(path string) (string, error) {
 			}
 			return maybeSources{maybeGitSource{url: u}}, nil
-		} else if ishg {
+		} else if !isgit {
```
We can just say `else` here since this covers the exact opposite case as the first block.

kevinburke commented on gps/deduce.go:
```diff
@@ -167,36 +169,74 @@ func (m bitbucketDeducer) deduceRoot(path string) (string, error) {
-			return nil, fmt.Errorf("%s is not a valid scheme for accessing either a git or hg repository", u.Scheme)
 		}
 
 		// No other choice, make an option for both git and hg
```
Can we flip the order here, I think git repos return quickly if you try to ping them with hg but the reverse is not true.

kevinburke commented:
@theckman what do you think? should we merge? what do we need to do?

theckman commented on gps/internal/bitbucketapi/bitbucketapi.go:
```diff
@@ -0,0 +1,142 @@
+	SCM string `json:"scm"`
+}
+
+const environmentVariable = "BITBUCKET_API_TOKEN"
```
@kevinburke I was re-reading through this, do you think we should namespace this? `DEP_BITBUCKET_API_TOKEN`?

kevinburke replied:
I think it's OK... I would like to make it as easy to guess as possible and the odds of someone needing two different tokens in the same environment seems low.

mvdan commented:
Dep was officially deprecated earlier this year, and the proposal to archive this repository [was accepted](https://github.com/golang/go/issues/38158). As such, I'm closing outstanding issues before archiving the repository. For any further comments, please use the proposal thread on the Go issue tracker. Thanks!
